File: src-tauri/src/get_installed_apps.rs
use tauri::api::process::Command;

#[tauri::command(async)]
pub async fn get_installed_apps() -> Vec<String> {
    let result = Command::new("powershell.exe")
        .args(["Get-StartApps | Format-Table -Wrap -Autosize"])
        .output()
        .expect("failed to execute process");

    // separate the stdout in lines

    let mut lines: Vec<String> = result.stdout.lines().map(|line| line.to_string()).collect();

    // remove first 6 lines

    for _ in 0..6 {
        lines.remove(0);
    }

    // filter empty lines

    lines = lines
        .into_iter()
        .filter(|line| line.len() > 0)
        .collect::<Vec<String>>();

    return lines;
}

#[tauri::command(async)]
pub async fn open_app_by_id(app_id: String) {
    let arg = "'shell:AppsFolder\\".to_string() + &app_id + "'";

    Command::new("powershell.exe")
        .args(["start", &arg])
        .output()
        .expect("failed to execute process");
}

// TODO - Use this to get the icon of the app, then do a match with the app name to get the icon
// $loc = Get-ChildItem HKLM:\Software\Microsoft\Windows\CurrentVersion\Uninstall
// $names = $loc |foreach-object {Get-ItemProperty $_.PsPath}
// foreach ($name in $names)
// {
//      Write-Host $name.Displayname
//      Write-Host $name.DisplayIcon
// }

File: app/src/extensions/core/apps-explorer/getInstalledApps.ts
import type { AppEntry } from './types'
import { invoke } from '@tauri-apps/api/core'

export const getInstalledApps = async () => {
  const rawInstalledApps = await invoke<string>('get_installed_apps')

  const lines = rawInstalledApps.split('\n').map(line => line.trim()).filter(line => line.length > 0)

  // lines 0, 2, 4... contains app name
  // lines 1, 3, 5... contains app id

  const apps: Record<string, AppEntry> = {}

  for (let i = 0; i < lines.length; i += 2) {
    const name = lines[i] as string
    const id = lines[i + 1] as string

    apps[name] = { name, id }
  }

  return apps;
}
